sympy/functions/special/bessel.py

```diff
@@ -677,64 +677,79 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         _, e = arg.as_coeff_exponent(x)
 
         if e.is_positive:
-            term_one = ((-1)**(nu -1)*log(z/2)*besseli(nu, z))
-            term_two = (z/2)**(-nu)*factorial(nu - 1)/2 if (nu).is_positive else S.Zero
-            term_three = (-1)**nu*(z/2)**nu/(2*factorial(nu))*(digamma(nu + 1) - S.EulerGamma)
-            arg = Add(*[term_one, term_two, term_three]).as_leading_term(x, logx=logx)
-            return arg
+            if nu is S.Zero:
+                #Equation 9.6.8 of Abramowitz and Stegun (10th ed, 1972).
+                term = -log(z) - S.EulerGamma + log(2)
+            else:
+                #Equation 9.6.9 of Abramowitz and Stegun (10th ed, 1972).
+                term = gamma(Abs(nu))*(z/2)**(-Abs(nu))/2
```

Comment:
I believe the series (`_eval_nseries`) implementation is correct. But I am a bit skeptical about this. Though the reference seems correct to me but can we have a series implementation where we extract the leading term out of all of them. We currently use https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/02/0008/ ( when `nu` is a natural number) but I think we need we can refer to this link (https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/01/) when `nu` is not integer. 
Could you have a look into this . 
We could have two series based implementations, one when `nu` is integer and one when `nu` is not equal to integer

Reply:
If we calculate the leading term for the two separate series expansions for integer and non-integer `nu` (basically just plug $k=0$ in both <https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/01/0003/> and <https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/02/0008/> and then choose the $z^{-\nu}$ term when $\nu\ne 0$), it does come out to be what I have currently implemented.

I could do the leading term calculation in Sympy (as was done earlier) instead of doing it manually and putting the result, but that just looks like it would complicate the code (introducing another pair of if-branches that end up with mathematically identical results) without improving readability or speed.

If you insist, I can make the change you suggest, but I just wanted to confirm before proceeding.

Reply:
Hmm, I would think on this and get back to you. I do think your approach would work equally fine but I would like to make sure we don't miss out on any cases which are symbolic in nature. Many times when we simplify things, we end up loosing on the symbolic aspect on things for eg we should also take care of any general `besselk(nu, z)` case where no numeric stuff is involved and it's all symbolic . Hence a substitution might look obvious to us numerically but we might miss out on things !
Actually why don't you go ahead and test out your code on some symbolic cases and series expansions.

Reply:
> Actually why don't you go ahead and test out your code on some symbolic cases and series expansions.

Do you mean something like the following?
```
assert besselk(n, x).as_leading_term(x).subs({n:S(5)/7}) == besselk(S(5)/7, x).series(x).as_leading_term(x)
assert besselk(n, x).as_leading_term(x).subs({n:3}) == besselk(3, x).series(x).as_leading_term(x)
```

These currently work, and I can add them as tests if required.

Reply:
I think so, yeah something like this should work (also for series I reckon) but if you say that series expansions don't work when the order is symbolic in nature we could keep it like that for now and take that up in some other pr

Reply:
I just checked the symbolic case. On master, we get wrong results:
```
In [6]: besselk(y, x).as_leading_term(x).subs({y:2})
Out[6]: 
   2             2    2              2       
  x ⋅log(x)   γ⋅x    x ⋅(3/2 - γ)   x ⋅log(2)
- ───────── - ──── + ──────────── + ─────────
      8        16         16            8    

In [7]: besselk(2, x).as_leading_term(x)
Out[7]: 
2 
──
 2
x 
```
(both the results above should've been the same; the second one is the correct answer)
With this PR, we get
```
In [1]: besselk(y, x).as_leading_term(x).subs({y:2})
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In[1], line 1
----> 1 besselk(y, x).as_leading_term(x).subs({y:2})

File /usr/lib/python3.12/site-packages/sympy/core/cache.py:72, in __cacheit.<locals>.func_wrapper.<locals>.wrapper(*args, **kwargs)
     69 @wraps(func)
     70 def wrapper(*args, **kwargs):
     71     try:
---> 72         retval = cfunc(*args, **kwargs)
     73     except TypeError as e:
     74         if not e.args or not e.args[0].startswith('unhashable type:'):

File /usr/lib/python3.12/site-packages/sympy/core/expr.py:3496, in Expr.as_leading_term(self, logx, cdir, *symbols)
   3494 if x not in self.free_symbols:
   3495     return self
-> 3496 obj = self._eval_as_leading_term(x, logx=logx, cdir=cdir)
   3497 if obj is not None:
   3498     from sympy.simplify.powsimp import powsimp

File /usr/lib/python3.12/site-packages/sympy/functions/special/bessel.py:708, in besselk._eval_as_leading_term(self, x, logx, cdir)
    706         term = gamma(Abs(nu))*(z/2)**(-Abs(nu))/2
    707     else:
--> 708         raise NotImplementedError(f"Cannot proceed without knowing if {nu} is zero or not.")
    710     return term.as_leading_term(x, logx=logx)
    711 elif e.is_negative:
    712     # Equation 9.7.2 of Abramowitz and Stegun (10th ed, 1972).

NotImplementedError: Cannot proceed without knowing if y is zero or not.
```

Reply:
I think that might do.